### PR TITLE
CAD-406 fixed space leak in LogBuffer

### PR DESCRIPTION
**Pull request by CodiePP**
description
-----------

- [x] there was a space leak in `LogBuffer`; this fix was possible with great input from @ksaric 

### testing

1.) compile using `stack --nix --profile`
2.) run using `stack --nix --profile exec example-complex -- +RTS -hc -L45 -RTS`
3.) analyse: `hp2ps -c example-complex.hp`

checklist
---------

- [x] compiles (`cabal new-clean; cabal new-build`)
- [x] tests run successfully (`cabal new-test`)
- [ ] documentation added and created (`cd docs; nix-shell --run make`)
- [x] link to an issue
- [ ] link to an epic
- [ ] add estimate points
- [x] add milestone (the same as the linked issue)

### CODE REVIEW

**Comment by CodiePP** [comment]
image

File: iohk-monitoring/src/Cardano/BM/Data/Counter.lhs [hunk 1]
```diff
@@ -69,7 +69,7 @@ nameCounter (Counter RTSStats           _ _) = "RTS"
 \subsubsection{CounterState}\label{code:CounterState}\index{CounterState}
 \begin{code}
 data CounterState = CounterState {
-      csCounters   :: [Counter]
+      csCounters   :: ![Counter]
```
Comment (mrBliss) [inline iohk-monitoring/src/Cardano/BM/Data/Counter.lhs]: Note that this just force the list to WHNF. Using `_` to indicate a thunk, this changes `csCounters = _` into `csCounters = _ : _`, so this is not forcing all the elements of the list.

Comment (ksaric) [inline iohk-monitoring/src/Cardano/BM/Data/Counter.lhs]: Not sure about this bang here.

**Comment by mrBliss** [comment]
Can you show a before and after? Otherwise, it seems like you're just adding bangs in random places.

File: iohk-monitoring/src/Cardano/BM/Data/Counter.lhs (hunk 1 above)
Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/Counter.lhs]: reverted it.

Comment (CodiePP) [reply iohk-monitoring/src/Cardano/BM/Data/Counter.lhs]: reverted it.

Comment (mrBliss) [reply iohk-monitoring/src/Cardano/BM/Data/Counter.lhs]: If this really is leaking space (we need some evidence for that first), then removing the bang is not the solution. Either use a stricter data structure than a list or force its contents (e.g., `Cardano.Prelude.forceElemsToWHNF`).

**Comment by mrBliss** [comment]
> Can you show a heap profile of before in addition to the one after? Otherwise, it seems like you're just adding bangs in random places.

@CodiePP informed me that the [JIRA issue](https://jira.iohk.io/browse/CAD-406) has a heap profile before and after.

Rather annoying that we have to start looking at multiple places now for information, instead of just GitHub, the place we are already using for issues, PRs, and code.

**Pull request merged**